rfcs/20190815-tfx-notebook.md

```diff
@@ -0,0 +1,563 @@
+:------------ | :-------
+**Author(s)** | Charles Chen (<a href="mailto:ccy@google.com">ccy@google.com</a>), Joe Lee (<a href="mailto:joeyounglee@google.com">joeyounglee@google.com</a>), Kenny Song (<a href="mailto:kennysong@google.com">kennysong@google.com</a>), Kevin Haas (<a href="mailto:khaas@google.com">khaas@google.com</a>), Pushkar Joshi (<a href="mailto:pushkarj@google.com">pushkarj@google.com</a>)
+**Sponsor**   | Konstantinos Katsiapis (katsiapis@google.com)
+**Updated**   | 2019-08-15
+
+## Objective
+
+We want to build a notebook user experience for modeling / iterative development
+using TFX Components. This will provide a fast, familiar environment for
+developing model and pipeline code with standard TensorFlow + TFX utilities,
+plus automatic notebook → pipeline export:
+
+*   Imperative,
+    [define-by-run](https://ai.googleblog.com/2017/10/eager-execution-imperative-define-by.html),
+    cell-by-cell workflow
+    *   Start directly from Notebook/Colab – no running pipeline needed
+    *   Run TFX components as you need them, in separate cells
+    *   No explicit DAG definitions or continuous execution
+*   Simple Python API per TFX component
+    *   ExampleGen, StatsGen, SchemaGen, Transform, Trainer, Evaluator
+    *   100% TFX compatible for automatic notebook → pipeline export
+*   Analyze artifacts natively in Notebook/Colab
+    *   Built-in TensorBoard, Facets, TFMA visualizations
+    *   Dataset, stats, eval metrics available in notebook for custom analysis
+*   Zero-setup, interactive onboarding tool for new TFX users on
+    [tensorflow.org](http://tensorflow.org)
+
+## Motivation
+
+The benefits of using a notebook include rapidly editing and running code,
+immediately seeing the execution and outputs of commands, and running quick
+one-off analyses in Python. It’s a simple, no-mental-overhead REPL environment
+for iterating on ideas.
+
+By combining the notebook experience + TFX components, users can easily run
+
+*   ExampleGen to generate the initial dataset used for training
+*   StatsGen to generate and visualize a statistical report of your data
+*   SchemaGen to generate a schema of your data (required input of Transform)
+*   Transform to write feature engineering strategies
+*   Trainer that wraps standard TF.Estimator or Keras code
+*   Evaluator to generate, slice, and visualize evaluation metrics
+*   Custom analyses on the output of any of these components with standard
+    Python
+
+To close the loop, the notebook will be automatically exported as a pipeline
+configuration that users can directly deploy as a scalable TFX pipeline. There
+is no additional modification required.
+
+## Target Users
+
+We target users who want to manually iterate on their models & components, and
+prefer a notebook environment for the benefits outlined above. This is a wide
+range of potential users, and from our user research, spans software engineers
+and data scientists within and outside of Google.
+
```

Comment:
It might be useful to think about these users more-- a lot of the below seems to be a replication of the existing APIs for use with colab, which implies the primary users will be going between the two systems/migrating from the existing. Is that the case, or are you imagining that the users of this will be entirely new?

Reply:
The proposed notebook API here centers around the TFX Component (https://www.tensorflow.org/tfx/guide#tfx_pipeline_components) which provides a set of functionality that the user may compose or extend (for example, [Kubeflow pipelines](https://www.kubeflow.org/docs/pipelines/overview/pipelines-overview/) is converging to be compatible and use the same interface).

Here, I believe that the TFX abstraction is at a higher level than the TF one, since it is concerned with composing / orchestrating / running these often-prebuilt components, whereas the TF one focuses more closely on the details of model computations.